text/0072-online-bulk-load-for-rawkv.md

```diff
@@ -0,0 +1,154 @@
+
+## Detailed design
+
+The solution proposed in this RFC is as follows:
+
+1. use distributed computing framework `Spark` to read the source files, convert
+and encode the trained data into Key Value format,
+2. sample the Key Value pairs, calculate the region split points, and call `split region and scatter` API
+3. repartition the Key-Value pairs RDD according to the new Regions boundaries, and sort the Key-Value pairs of each partition,
+use the [ImportSST.RawWrite](#proto) to concurrently send the Key-Value pairs to the TiKV server and generate SST files on the TiKV server,
+4. use the [Ingest API](https://github.com/pingcap/kvproto/blob/release-5.0/proto/import_sstpb.proto#L53)
+to ingest the generated SST files into TiKV.
+
+![Bulk Load](../media/bulk-load.png)
+
+###  Why `split region and scatter`
+
+There may be a situation that all the data to load belongs to one Region. In such case only one tikv-server instance will be used to process the bulk load requests. Before doing bulk load, we need to sample the Key Value pairs, calculate the region split points, and call `split region and scatter` API, so that most of the tikv-server instances can be used to process the bulk load requests.
+
+### Proto
+
+```
+service ImportSST {
+  rpc RawWrite(stream RawWriteRequest) returns (RawWriteResponse) {}
+}
+
+message RawWriteBatch {
+    uint64 ttl = 1;
+    repeated Pair pairs = 2;
+}
+
+message RawWriteRequest {
+    oneof chunk {
+        SSTMeta meta = 1;
+        RawWriteBatch batch = 2;
+    }
+}
+
+message RawWriteResponse {
+    Error error = 1;
+    repeated SSTMeta metas = 2;
+}
+```
+
+### Rate Limit
+
+We want to control the impact to the online serving (P99 latency) during the bulk load, so `Rate Limit` is an important component.
+
+#### config import.num-threads
+
+First of all, users can config the thread number of the import module, which is already implemented.
+
+```
+[import]
+## Number of threads to handle RPC requests.
+# num-threads = 8
+```
+
+#### config storage.io-rate-limit
+
+[storage.io-rate-limit](https://docs.pingcap.com/tidb/stable/tikv-configuration-file#storageio-rate-limit) limits the maximum I/O bytes that a server can write to or read from the disk.
+
+Users can use `io-rate-limit` to limit Disk Write of the `import` module:
+
+- config storage.io-rate-limit.max-bytes-per-sec
+- config storage.io-rate-limit.import-priority = "low"
+
+
+If the total disk write < max-bytes-per-sec, the `import` module will not be limited.
+If the total disk write >= max-bytes-per-sec, the `import` module will be limited.
+
+
+### How to measure the performance impact to the online services
+
+One target of the bulk load proposal is that the impact to the online serving (P99 latency) should be under control. The performance impact to the online services can be mesasured as follows:
+
+1. Run stress testing program to simulate the business logic, in which `Raw Get`, `Raw Put`, `Raw Batch Put`... will be called continuously.
```

Comment:
is there any YCSB workload to satisify the requirement?

Reply:
Users can choose one from WorkloadA to WorkloadF according to their workload.
https://tikv.org/docs/5.1/deploy/performance/instructions/#step-3-run-a-ycsb-workload
If it does not satisfy, users can implement their own workload.

Reply:
the main purpose of this section is to guide the performance test ourselves. we can directly use YCSB in this step.

Reply:
sure. I suggest to use `Workload B` (Read mostly workload), which is similar to  the `Feature Store Online Serving` case.